bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -76,339 +67,329 @@
-    // which relies on a ThreadLocal.
-    this.isAppEngine = credentials.getClass().getName().contains("AppEngineCredentials");
+
+    // From MoreExecutors
+    this.isAppEngine = System.getProperty("com.google.appengine.runtime.environment") != null;
+
+    rateLimiter = RateLimiter.create(1);
   }
 
-  /** {@inheritDoc} */
+  /**
+   * {@inheritDoc}
+   */
   @Override
   public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(MethodDescriptor<ReqT, RespT> method,
       CallOptions callOptions, Channel next) {
-    return new CheckedForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
+    return new SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
       @Override
-      protected void checkedStart(Listener<RespT> responseListener, Metadata headers)
-          throws StatusException {
-        headers.put(AUTHORIZATION_HEADER_KEY, getHeader());
-        delegate().start(responseListener, headers);
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        HeaderCacheElement headerCache = getHeaderSafe();
+
+        if (!headerCache.status.isOk()) {
+          responseListener.onClose(headerCache.status, new Metadata());
+          return;
+        }
+
+        headers.put(AUTHORIZATION_HEADER_KEY, headerCache.header);
+
+        delegate().start(new UnauthResponseListener<>(responseListener, headerCache), headers);
       }
     };
   }
 
-  /**
-   * Refreshes the OAuth2 token asynchronously. This method will only start an async refresh if
-   * there isn't a currently running asynchronous refresh or the credentials are for App Engine.
-   * App Engine has some credentials related state in a {@link ThreadLocal} which prevents it from
-   * running asynchronously. In the App Engine case, this method will defer to
-   * {@link #syncRefresh()}.
-   *
-   * @throws IOException
-   */
-  public void asyncRefresh() throws IOException {
-    if (isAppEngine) {
-      syncRefresh();
-    } else if (canRefresh()) {
-      executor.execute(new Runnable() {
-        @Override
-        public void run() {
-          doRefresh();
-        }
-      });
+  private HeaderCacheElement getHeaderSafe() {
+    try {
+      return getHeader();
+    } catch (Exception e) {
+      return new HeaderCacheElement(
+          Status.UNAUTHENTICATED
+              .withDescription("Unexpected failure get auth token")
+              .withCause(e)
+      );
     }
   }
 
-  private boolean canRefresh() {
-    return !isRefreshing.get() && getCacheState(this.headerCache.get()) != CacheState.Good;
-  }
-
   /**
-   * <p>syncRefresh.</p>
-   *
-   * @throws java.io.IOException if any.
+   * Get the http credential header we need from a new oauth2 AccessToken.
    */
-  public void syncRefresh() throws IOException {
-    synchronized (isRefreshing) {
-      if (!isRefreshing.get()) {
-        doRefresh();
+  private HeaderCacheElement getHeader()
```

Comment:
Why do we need to incur the lock penalty when CacheState == CacheState.Good?  Can we optimize for that situation?